New search module.

msmorgan opened the pull request.

msmorgan commented:
@mendess I'll continue working on this later, but if you care to take a look at `src/search.rs`, I'd appreciate some early feedback.

mendess commented on src/search.rs:
```diff
@@ -0,0 +1,755 @@
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct Compare<T> {
+    op: Option<CompareOp>,
```
why is the compare operation optional?

msmorgan replied:
Otherwise it's a colon.

mendess replied:
I don't get ti :thinking:

msmorgan replied:
I had two other choices. I could make the `Param::Value` look like `Param::Value(ValueKind, Option<CompareOp>, Box<...>)` (where `None` is a colon, such as in the param `name:"stormcrow"`), or I could have added a third variant `Param::Compare(ValueKind, CompareOp, Box<...>)`. I wanted to be able to write `cmc(gt(3))` and such,  so I figured I would use a structure for comparisons, and then convert bare values to compare without an `op` (i.e., a colon).

The Scryfall API might treat : and = equally. With some testing I could remove the colon case entirely if that turns out to be true.

jheyens replied:
> The Scryfall API might treat : and = equally. With some testing I could remove the colon case entirely if that turns out to be true.

No, Scryfall does not treat them equally.
Example: 
https://scryfall.com/search?q=id%3Aubg&unique=cards&as=grid&order=name
vs
https://scryfall.com/search?q=id%3Dubg&unique=cards&as=grid&order=name

msmorgan replied:
@jheyens Thank you! I just discovered this myself. It seems like `:` is equivalent to `>=` when it comes to colors.

mendess reviewed (commented):
I'm trying to follow the equivalences in functionality from the previous implementation. I can see that you've generalized the idea that some parameters take parameters but I think you should also distinguish those that can be compared from the ones that can't? I'm not sure how much work that would be but it's a bit confusing that `Compare<T>` only compares sometimes.

Other than that I think it's looking really good so far.

msmorgan commented on src/search.rs:
```diff
@@ -0,0 +1,755 @@
+                },
+                expr
+            )?;
+        }
```
But that method has such a terrible name!

mendess replied:
hahaha, yeah kinda. Sadly the good name one (`join`) is not as good. And I think the name isn't that bad, it formats an iterator and has too modes, use the default `Display` impl or use a custom one (`format_with`)

msmorgan replied:
TBH I didn't know itertools had that method, so I'll use it instead of reinventing the wheel. Saves like 10 lines.

mendess commented:
Is the `Rc` in `ParamImpl` so you can derive `Clone`?

msmorgan commented:
> Is the `Rc` in `ParamImpl` so you can derive `Clone`?

Correct. The `Clone` trait is not object-safe, but it seems useful to be able to clone params. Do you have another idea?

msmorgan commented:
I'm not so how necessary it is to use `VecDeque`. There's an edge case where it's faster, but it'll be pretty infrequently encountered, since queries will likely be constructed from start to end more often than not. It would be convenient if you could write: `Query::And(vec![q1, q2])` instead of `Query::And(vec![q1, q1].into_iter().collect())`, and to me that outweighs the benefits of `VecDeque`.

Another option would be not to collapse the `And` vector when it's `Query::Param(..)and(Query::And(..))`. In that case it would output, `(q1 AND (q2 AND q3 AND q4))`, which is still a valid scryfall query.

I think I'm going to get rid of the operator overloads too. They're neat but they're making me a little queasy.

mendess commented:
You could make the enum have a private impl thing and make a function to add that takes a vec or any other IntoIter thing? But I don't think it's worth the trouble and complexity.

I agree that changing to vec is the right move.

msmorgan commented:
@mendess @jheyens

I'd like your opinion about the names of color parameters.

Some background: in scryfall search syntax, you can use the parameter `artists` to search for a printing with a certain number of contributing artists (e.g. [cards with 2 artists](https://scryfall.com/search?q=artists%3D2&unique=cards&as=grid&order=name)), and a card with a certain number of illustrations (e.g. [cards with at least 10 unique art](https://scryfall.com/search?q=illustrations%3E%3D10)). I've chosen to rename these for this crate, into `artist_count` and `illustration_count` since I believe those are clearer and more discoverable.

However, a small problem arises with the `color` parameter. You can specify either a set of colors/alias (e.g. [simic cards](https://scryfall.com/search?q=c%3Dsimic)) or a number of colors (e.g. [tri-color cards from Khans of Tarkir](https://scryfall.com/search?q=set%3Aktk+AND+color%3A3)). I'm considering splitting these into separate methods `color` and `color_count` (as well as `identity` and `identity_count`), so there's one function that takes a set of colors and another that takes a number. Do you think I should split it up like this?

If you prefer, you could also veto my idea to rename `artist_count` and `illustration_count`, changing it back to `artists` and `illustrations` to match the scryfall syntax more closely. In that case, I'd use a single function for `color` and `identity`.

mendess commented:
Hmm... I've been reading and re reading the code. I think having explicit methods for counts is good.

About the color aliases thing, we could also ignore the scryfall shorts kinda
```rs
const SIMIC: Colors = [Color::Blue, Color::Green].into();
```
(this probably won't compile but it's just to get the idea across, and also makes me thing the `Colors` struct should have better ways to construct :thinking:)

This way we simplify things on our end without losing the confy scryfall setup.

msmorgan commented:
I've overloaded `ops::BitOr` because `Colors` works like bitflags, but it might also be nice to overload `ops::Sub` to remove colors. This makes me wonder if `ops::Add` is a good idea instead of `ops::BitOr`.

mendess commented on src/card/color.rs:
```diff
@@ -184,12 +182,45 @@ impl Colors {
+
+    /// Produces a new instance with the specified color removed.
+    pub const fn without(self, color: Color) -> Self {
+        Colors(self.0 & !(color as u8))
```
Isn't this wrong?
won't this set the highest 3 bits?

msmorgan replied:
No, I don't think so. `!(color as u8)` will have its highest 3 bits set, but `self.0` never will, so since the operation is `&` this should be correct.

msmorgan commented:
I could totally use `String` instead of `Rc<dyn ParamValue>` in the `Param::Value` variant and get `Clone` for free. Why didn't I think of this before? -_-

mendess commented:
Hmm.. I didn't originally use `Vec<String>` because I thought it was a good idea to do it lazily. But in retrospect it's a bit retarded.

Would it be possible to use a single `String` and serialise as you take params?

Cows might be a bit overkill no?

msmorgan commented:
What I meant was
```
enum ParamImpl {
    Value(ValueKind, Option<CompareOp>, String),
}
```

I'd rather the `Query` objects be an expression tree, since that makes `Debug` output clearer and allows for collapsing statements in the `Query::and` and `Query::or` methods.

P.S. Be careful using that R-word. It's probably against the community guidelines.

mendess commented:
That makes sense. If this ever becomes a performance problem (which it probably won't) we could change it or something. I think it's good.

msmorgan commented:
I'm starting to wonder if the division between `Param::Property` and `Param::Value` is necessary, now that values are stored as Strings. Then we'd use `is_first_print(false)` instead of `not(Property::IsFirstPrint)`...

Or I could make `Property` private and operate as it currently does, and create Query-returning functions for each type of property. They could take a boolean argument for convenience and clarity.

mendess commented:
I don't know if I follow the second part about query returning methods.

Having functions over types is always an interesting duality, because types you can pass around, functions not so much. I don't know what I would pick here. The function is more legible tho.

msmorgan commented:
I converted them to free functions but I don't think I like it. I've come up with several variations on the API for these sorts of parameters. Below are some examples, using [this query](https://scryfall.com/search?q=is%3Aphyrexian+new%3Aart+-is%3Afirstprint+o%3A%22%7BT%7D%22) and [this query](https://scryfall.com/search?q=is%3Amanland+new%3Aart+-is%3Afirstprint+-is%3Adigital+is%3Apromo). What do you think?

## Free functions

The state as of the [current commit](https://github.com/mendess/scryfall-rs/pull/22/commits/87f8bb819ecba4ffe2cf53a8e498dac002426677).

```rust
let birthing_pod = Query::And(vec![
    is_phyrexian(true),
    new_art(true),
    is_first_print(false),
    oracle_text("{T}"),
])
.random()?;

let celestial_colonnade = is_creature_land(true)
    .and(new_art(true))
    .and(is_first_print(false))
    .and(is_digital(false))
    .and(is_promo(true))
    .random()?;
```

## Criterion enum with `is` function

```rust
let birthing_pod = Query::And(vec![
    is(Criterion::Phyrexian),
    is(Criterion::NewArt),
    not(Criterion::FirstPrint),
    oracle_text("{T}"),
])
.random()?;

let celestial_colonnade = is(Criterion::CreatureLand)
    .and(is(Criterion::NewArt))
    .and(not(Criterion::FirstPrint))
    .and(not(Criterion::Digital))
    .and(Criterion::Promo) // As an argument to `and`, `is` is optional since `Criterion: Into<Query>`.
    .random()?;
```

# Multiple enums

The idea here is to split parts of `Criterion` into separate categories, such as:

- `CardCriterion`  - for criteria that are the same for all printings of a card
- `PrintCriterion`  - for criteria that apply to specific printings
- `ProductType` - a.k.a. `SoldIn`, this works as both a card criterion with `in:` and a print criterion with `is:`. This should perhaps should be removed in favor of `SetType` and implemented in the `value` module.
- `LandFamily` - for card criteria with groupings of lands, e.g. `is:fetchland`.

```rust
let birthing_pod = Query::And(vec![
    is(CardCriterion::Phyrexian),
    is(PrintCriterion::NewArt),
    not(PrintCriterion::FirstPrint),
    oracle_text("{T}"),
])
.random()?;

let celestial_colonnade = is(LandFamily::CreatureLand)
    .and(is(PrintCriterion::NewArt))
    .and(not(PrintCriterion::FirstPrint))
    .and(not(PrintCriterion::Digital))
    .and(PrintCriterion::Promo) // As an argument to `and`, `is` is optional since `Criterion: Into<Query>`.
    .random()?;
```

mendess commented:
I'm not sure which I like best. I kinda wish there was a shorter word instead of `Criterion` but maybe that's a bit nit picky.

It's a bit hard to decide what is more ergonomic without seeing usecases.
- Functions read nicer, aesthetically but besides that I don't see an advantage.
- Is it likely people will want to pass/store these enums? If yes then it's better than functions.
- In that case is it likely people will want to treat all `Criterion` the same? If yes then having categories like this might become cumbersome.

Also, isn't land family a special case of `CardCriterion`? Not a criticism just trying to see if I understand the categories.

msmorgan commented:
> I kinda wish there was a shorter word instead of `Criterion`

Me too. `Property` was too generic and `Criterion` is just one additional letter and matches the wording on the advanced search page. The other word I had in mind is `Flag` but that doesn't seem quite right.

> * Is it likely people will want to pass/store these enums? If yes then it's better than functions.

Maybe, but they could always store the result of calling the functions. The output of every one of these functions is a `Query`.

> Also, isn't land family a special case of `CardCriterion`?

Yup, that's true. They're a semantic fit for `CardCriterion`, but I thought creating a separate enum would be more appropriate since there are so many in this class.

Unfortunately I can't call the enums simply `Card` and `Printing`, despite `Card::IsPhyrexian` and `Printing::IsDigital` read quite nicely. The type `Card` already exists. If I wanted to be clever, I guess I could use the names `CardIs` and `PrintingIs` (by the way, on Scryfall `has:` and `is:` are aliases of one another), and get `CardIs::Phyrexian` and `PrintingIs::Digital`.

One of the benefits of having the card and printing criteria in separate enums is it may enable more advanced query processing.

mendess commented:
I agree on `Flag` being bad and non descriptive.

I agree with the `LandCriterion` category, lands are so relevant in magic and filtering them too.

I actually like `CardIs`, despite being a really unorthodox way to name an enum I think the benefits in readability are really good! I would go with that.

msmorgan commented:
Ok, I have one unanswered question if I'm going to go with `CardIs`: how do I convert it into a `Query`? `not(CardIs::Phyrexian)` works, as does `CardIs::Phyrexian.into()`. But now the `is` function name (`is(CardIs::Phyrexian)`?) doesn't make sense. I could name it `criterion` instead, but that would just be kicking the can down the road. Do you have any ideas? Perhaps `q`?

msmorgan commented:
I think the [`criteria`](https://github.com/mendess/scryfall-rs/pull/22/files#diff-3c4e8e0097dfc7f84aa4f16757501f3041760d100cbf74225d690baf6f3aa157) module is done now. I'm pretty happy with how it turned out. Thank you @mendess for suggesting I go with `CardIs` and `PrintingIs`, as they read quite nicely. For now, I decided against adding a free function to convert to `Query`.

mendess commented:
Looks good, and the from impl should be good. Sorry I didn't reply to the last message, I've been really busy!

You're having some really good ideas here. Keep it up!!

msmorgan commented:
I had to put this on hold for a bit, but I just finished up. I'll go ahead and unmark it as a draft.

mendess reviewed (approved):
Coming back to this after a while I didn't remember a lot so I think this makes it a better review test haha.

First of all thank you and congrats! I think you did a great job here!

On the inside it did become quite complex, but that's to be expected and I could follow it by reading the docs for a bit so I think it passes the intuitive part, it could have a bit more code examples, but at the end of the day a user doesn't really need to understand much more than how they can compose query parameters and after doing it once it should be obvious how to go from there to make more, much like other good rust apis, you don't need to understand it, the compiler will check it and guide you through it better than any docs.

Overall well done!

msmorgan commented:
What would you think of doing a merge commit instead of a squash?
